### PR TITLE
Anisotropy cookbook

### PR DESCRIPTION
**Pull request by KiralyAgi**
This is a start for a cookbook for Rayleigh-Taylor instabilities using anisotropic viscosity terms.

### CODE REVIEW

**Comment by bangerth** [comment]
You just said that you have further changes for this. You can just push them here, but talk to me if you have questions of how to do that!

File: cookbooks/anisotropic_viscosity/anisotropic_viscosity.prm [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# This tests an implementation of anisotropic viscosity terms in the Stokes
+# equation. In this test we create new assembler
+# classes in anisotropic_viscosity.cc that extend the standard assemblers
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/anisotropic_viscosity.prm]:
```suggestion
# class in anisotropic_viscosity.cc that extends the standard assemblers
```

File: cookbooks/anisotropic_viscosity/anisotropic_viscosity.prm [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+# with a fourth order viscosity tensor. The chosen material model 'anisotropic'
+# then replaces the standard assemblers with the custom ones during the call
+# to the set_assemblers signal, and during every call to evaluate()
+# fills the AdditionalMaterialOutputs object of type AnisotropicViscosity with
+# the chosen viscosity tensor that is selected in this input file.
+# This way the solved equations are changed without modifying the main codei
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/anisotropic_viscosity.prm]:
```suggestion
# This way the solved equations are changed without modifying the main code
```

File: cookbooks/anisotropic_viscosity/anisotropic_viscosity.prm [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+end
+
+subsection Mesh refinement
+  set Initial adaptive refinement        = 0
+  set Initial global refinement          = 5
+end
+
+# The parameters below this comment were created by the update script
+# as replacement for the old 'Model settings' subsection. They can be
+# safely merged with any existing subsections with the same name.
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/anisotropic_viscosity.prm]: I think you can safely remove these lines :-)

File: source/postprocess/visualization/viscosity_tensor.cp [hunk 4]
```diff
@@ -0,0 +1,147 @@
+/*
```
Comment (bangerth) [inline source/postprocess/visualization/viscosity_tensor.cp]: you want to use the extension `.cc` for this file, not `.cp`.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 5]
```diff
@@ -0,0 +1,858 @@
+#include <aspect/introspection.h>
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: can you please copy the copyright header here that is present in all other files?

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 6]
```diff
@@ -0,0 +1,858 @@
...
+#include <deal.II/grid/tria_iterator_base.h>
+#include <array>
+#include <cmath>
+#include <functional>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#ifndef __aspect__av_material_h
+#define __aspect__av_material_h
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: move this above the `#include` statements above.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 7]
```diff
@@ -0,0 +1,858 @@
...
+  {
+    using namespace dealii;
+
+              /**
+               * Additional output fields for anisotropic viscosities to be added to
+               * the MaterialModel::MaterialModelOutputs structure and filled in the
+               * MaterialModel::Interface::evaluate() function.
+               */
+              template <int dim>
+              class AnisotropicViscosity : public NamedAdditionalMaterialOutputs<dim>
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: I think you haven't run the indentation script, or did the script simply not pick up your file?

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 8]
```diff
@@ -0,0 +1,858 @@
...
+                      }
+                          return names;
+                    }
+                  }
+
+              template <int dim>
+              AnisotropicViscosity<dim>::AnisotropicViscosity (const unsigned int n_points)
+              :
+
+              NamedAdditionalMaterialOutputs<dim>(make_AnisotropicViscosity_additional_outputs_names<dim>()),
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: remove the empty line

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 9]
```diff
@@ -0,0 +1,858 @@
...
+}
+
+#endif
+
+namespace aspect
+{
+
+
+    
+    namespace Assemblers
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: remove the empty lines

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 10]
```diff
@@ -0,0 +1,858 @@
...
+
+        template <int dim>
+        void
+        StokesPreconditionerAnisotropicViscosity<dim>::
+        execute (internal::Assembly::Scratch::ScratchBase<dim>   &scratch_base,
+                 internal::Assembly::CopyData::CopyDataBase<dim> &data_base) const
+        {
+            internal::Assembly::Scratch::StokesPreconditioner<dim> &scratch = dynamic_cast<internal::Assembly::Scratch::StokesPreconditioner<dim>& > (scratch_base);
+            internal::Assembly::CopyData::StokesPreconditioner<dim> &data = dynamic_cast<internal::Assembly::CopyData::StokesPreconditioner<dim>& > (data_base);
+            //std::cout<<"StokesPreconditioner was called";
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: remove the commented out line

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 11]
```diff
@@ -0,0 +1,858 @@
...
+                                                                                .extractors.pressure].value(i, q);
+                        ++i_stokes;
+                    }
+                    ++i;
+                }
+                
+                const double eta = scratch.material_model_outputs.viscosities[q];
+                const double one_over_eta = 1. / eta;
+                
+                const bool use_tensor = (anisotropic_viscosity != nullptr);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: Should that not always be the case for you? If so, you could do
```
  Assert (anisotropic_viscosity != null_ptr, ...);
```
and replace `use_tensor` by `true` in everything below.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 12]
```diff
@@ -0,0 +1,858 @@
...
+        void
+        StokesPreconditionerAnisotropicViscosity<dim>::
+        create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
+        {
+            const unsigned int n_points = outputs.viscosities.size();
+            
+            if (outputs.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >() == nullptr)
+            {
+                outputs.additional_outputs.push_back(
+                                                     std::make_shared<MaterialModel::AnisotropicViscosity<dim>> (n_points));
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: I *think* that in the current development version, this should be `std_cxx14::make_unique`.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 13]
```diff
@@ -0,0 +1,858 @@
...
+                
+                
+                // Viscosity scalar
+                const double eta = (scratch.rebuild_stokes_matrix
+                                    ?
+                                    scratch.material_model_outputs.viscosities[q]
+                                    :
+                                    numbers::signaling_nan<double>());
+                
+                const bool use_tensor = (anisotropic_viscosity != nullptr);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: same comments in this function as in the previous one

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 14]
```diff
@@ -0,0 +1,858 @@
...
+        void
+        StokesIncompressibleTermsAnisotropicViscosity<dim>::
+        create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
+        {
+            const unsigned int n_points = outputs.viscosities.size();
+            
+            if (outputs.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >() == nullptr)
+            {
+                outputs.additional_outputs.push_back(
+                                                     std::make_shared<MaterialModel::AnisotropicViscosity<dim>> (n_points));
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: same as above; there are a couple more places below

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 15]
```diff
@@ -0,0 +1,858 @@
...
+        void
+        ShearHeatingAnisotropicViscosity<dim>::
+        create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &material_model_outputs) const
+        {
+            const unsigned int n_points = material_model_outputs.viscosities.size();
+            
+            if (material_model_outputs.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >() == nullptr)
+            {
+                material_model_outputs.additional_outputs.push_back(
+                                                                    std::make_shared<MaterialModel::AnisotropicViscosity<dim>> (n_points));
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: and here

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 16]
```diff
@@ -0,0 +1,858 @@
...
+            }
+            
+            this->get_material_model().create_additional_named_outputs(material_model_outputs);
+        }
+    }
+    
+    namespace MaterialModel
+    {
+        using namespace dealii;
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: do you actually need this?

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 17]
```diff
@@ -0,0 +1,858 @@
...
+            is_compressible () const;
+            
+        private:
+            void set_assemblers(const SimulatorAccess<dim> &,
+                                Assemblers::Manager<dim> &assemblers) const;
+            
+            // Constitutive tensor
+            SymmetricTensor<4,dim> C;
+        };
+        template <int dim>
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: add three empty lines between functions/classes

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 18]
```diff
@@ -0,0 +1,858 @@
...
+        class AV : public MaterialModel::Simple<dim>
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: Can you add comments in front of this class and the previous one that explain what these classes do?

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 19]
```diff
@@ -0,0 +1,858 @@
...
+                virtual void evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                                       MaterialModel::MaterialModelOutputs<dim> &out) const;
+                static void declare_parameters (ParameterHandler &prm);
+                virtual void parse_parameters (ParameterHandler &prm);
+                virtual bool is_compressible () const;
+                virtual double reference_viscosity () const;
+                virtual double reference_density () const;
+                virtual void create_additional_named_outputs(MaterialModel::MaterialModelOutputs<dim> &out) const;
+              private:
+                double eta_N, eta_0;
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: can you document what these variables are? I think that your `eta_0` is the ratio of viscosities, but then `eta` is maybe not such a good choice for the symbol/variable name. I would call it `viscosity_ratio`.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 20]
```diff
@@ -0,0 +1,858 @@
...
+                static double delta (const unsigned int i, const unsigned int j);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: can you document what this function does?

Comment (bangerth) [reply cookbooks/anisotropic_viscosity/av_material.cc]: Also, I would let it return an integer

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 21]
```diff
@@ -0,0 +1,858 @@
...
+	void
+	AV<dim>::
+	initialize()
+		{
+		this->get_signals().set_assemblers.connect (std::bind(&AV<dim>::set_assemblers,
+                                                          std::cref(*this),
+                                                          std::placeholders::_1,
+                                                          std::placeholders::_2));
+		AssertThrow((dim==2),
+		                ExcMessage("For now this works only in 2D"));
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]:
```suggestion
		                ExcMessage("For now the anisotropic viscosity assemblers work only in 2D."));
```

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 22]
```diff
@@ -0,0 +1,858 @@
...
+
+		}
+	template <int dim>
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: three empty lines between functions

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 23]
```diff
@@ -0,0 +1,858 @@
...
+	void
+	AV<dim>::evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                      MaterialModel::MaterialModelOutputs<dim> &out) const
+	{
+
+      MaterialModel::AnisotropicViscosity<dim> *anisotropic_viscosity =
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: no empty line here

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 24]
```diff
@@ -0,0 +1,858 @@
...
+      c_idx_n.push_back (this->introspection().compositional_index_for_name("nk"));
+
+    // Get the grad_u tensor, at the center of this cell, if possible.
+
+    std::vector<Tensor<2,dim> > velocity_gradients (in.position.size());
+    if (in.current_cell.state() == IteratorState::valid)
+      {
+    	std::vector<Point<dim> > quadrature_positions(in.position.size());
+    	          for (unsigned int i=0; i < in.position.size(); ++i)
+    	            quadrature_positions[i] = this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: This procedure is actually quite inefficient. There are easier ways to compute all of this. I suspect that you haven't done this yourself but copied it from somewhere. If you let me know where that is, then I can fix it there and point you at it.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 25]
```diff
@@ -0,0 +1,858 @@
...
+        // n needs to be a nonzero vector
+
+        Tensor<1,dim> n_new;
+        n_new = n + (n_dot * this->get_timestep());
+        Assert (n_new.norm() != 0, ExcInternalError());
+        n_new /= n_new.norm();
+        if (this->get_timestep()==0) //because at the start of the model timestep is 0
+        {
+
+        	n_dot=(n/n.norm())-n;
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: no need for the empty line

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 26]
```diff
@@ -0,0 +1,858 @@
...
+              for (unsigned int j=0; j<dim; ++j)
+                for (unsigned int k=0; k<dim; ++k)
+                  for (unsigned int l=0; l<dim; ++l)
+                    Lambda[i][j][k][l] = 1./2. * (n[i]*n[k]*delta(l,j)
+                                                  + n[j]*n[k]*delta(i,l)
+                                                  + n[i]*n[l]*delta(k,j)
+                                                  + n[j]*n[l]*delta(i,k))
+                                         - 2*n[i]*n[j]*n[k]*n[l];
+
+              if (anisotropic_viscosity != nullptr)
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: Should that not always be true?

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 27]
```diff
@@ -0,0 +1,858 @@
...
+    	                                   update_gradients);
+    	fe_values.reinit (in.current_cell);
+
+
+        fe_values[this->introspection().extractors.velocities]
+        .get_function_gradients(this->get_solution(), velocity_gradients);
+
+      }
+
+    std::vector<std::array<double,3> > Viscoeigenvalues(in.position.size()); //This is good only in 2D
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: remove this line

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 28]
```diff
@@ -0,0 +1,858 @@
...
+                  ViscoTensor[0][1]=anisotropic_viscosity->stress_strain_directors[q][0][0][1][1];
+                  ViscoTensor[0][2]=anisotropic_viscosity->stress_strain_directors[q][0][0][0][1] * sqrt(2);
+                  ViscoTensor[1][0]=anisotropic_viscosity->stress_strain_directors[q][1][1][0][0];
+                  ViscoTensor[1][1]=anisotropic_viscosity->stress_strain_directors[q][1][1][1][1];
+                  ViscoTensor[1][2]=anisotropic_viscosity->stress_strain_directors[q][1][1][0][1] * sqrt(2);
+                  ViscoTensor[2][0]=anisotropic_viscosity->stress_strain_directors[q][0][1][0][0] * sqrt(2);
+                  ViscoTensor[2][1]=anisotropic_viscosity->stress_strain_directors[q][0][1][1][1] * sqrt(2);
+                  ViscoTensor[2][2]=anisotropic_viscosity->stress_strain_directors[q][0][1][0][1] * 2;
+
+                  Viscoeigenvalues[q] = eigenvalues(ViscoTensor);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: You don't actually need an array of these eigenvalues -- you really only need them one at a time. So declare
```
  const std::array<double,3> visco_eigenvalues = eigenvalues(ViscoTensor);
```
here. Then you don't need the array declaration above (as commented on).

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 29]
```diff
@@ -0,0 +1,858 @@
...
+                  for (unsigned int i=0; i<3; ++i)
+                  {
+                	  if (Viscoeigenvalues[q][i]<0)
+                		  cout<<" eig1 "<<Viscoeigenvalues[q][0] <<" eig2 "<<Viscoeigenvalues[q][1] <<" eig3 "<<Viscoeigenvalues[q][2] <<" n "<<n << std::endl;
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: remove the debug output. the assertion below is good enough.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 30]
```diff
@@ -0,0 +1,858 @@
...
+                	  AssertThrow((Viscoeigenvalues[q][i]>0),
+                	  		                ExcMessage("Error! Negative eigenvalue in the viscosity tensor"));
+                  //Assert (anisotropic_viscosity->stress_strain_directors[q] has only positive eigenvalues);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: remove the comment

File: include/aspect/postprocess/visualization/viscosity_tensor.h [hunk 31]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (bangerth) [inline include/aspect/postprocess/visualization/viscosity_tensor.h]: Is this file a part of the coookbook, or do you want it to be usable in a broader context? I believe it's the former, and in that case you should probably move this .h and the corresponding .cc file into the cookbook folder.

File: include/aspect/postprocess/visualization/viscosity_tensor.h [hunk 32]
```diff
@@ -0,0 +1,93 @@
...
+       * - \tfrac 13 \textrm{trace}\ \varepsilon(\mathbf u) \mathbf 1) +pI =
+       * 2\eta (\varepsilon(\mathbf u) - \frac 13 (\nabla \cdot \mathbf u)
+       * \mathbf I) + pI$.  The second term in the parentheses is zero if the
+       * model is incompressible.
+       *
+       * The member functions are all implementations of those declared in the
+       * base class. See there for their meaning.
+       */
+      template <int dim>
+      class Stress
```
Comment (bangerth) [inline include/aspect/postprocess/visualization/viscosity_tensor.h]: The file is called `visualization/viscosity_tensor.h`, but you declare a `Stress` class. Is that a mistake?

File: source/postprocess/visualization/viscosity_tensor.cp [hunk 33]
```diff
@@ -0,0 +1,147 @@
...
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      template <int dim>
+      void
+      Stress<dim>::
+      evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
```
Comment (bangerth) [inline source/postprocess/visualization/viscosity_tensor.cp]: same here?

File: cookbooks/anisotropic_viscosity/av_material.cc (hunk 24 above)
Comment (KiralyAgi) [reply cookbooks/anisotropic_viscosity/av_material.cc]: It was copied from visco_plastic.cc (line 460). Based on a quick search, it is also in the viscoelastic.cc

**Comment by KiralyAgi** [comment]
I corrected the files.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 34]
```diff
@@ -0,0 +1,885 @@
...
+
+
+          fe_values[this->introspection().extractors.velocities]
+          .get_function_gradients(this->get_solution(), velocity_gradients);
+
+        }
+
+      for (unsigned int q=0; q<in.position.size(); ++q)
+        {
+          out.densities[q] = (in.composition[q][c_idx_gamma] > 0.8 ? 1 : 0);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: Why do you need to set densities etc here and in the following lines? Don't you inherit these values from the `Simple` model that you call above via `Simple<dim>::evaluate(in, out)`?

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 35]
```diff
@@ -0,0 +1,885 @@
...
+                  n_dot=(n/n.norm())-n;
+                }
+              else
+                {
+                  n_dot = (n_new-n)/ this->get_timestep();
+                }
+            }
+          else
+            {
+              n_dot=0;
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]:
```suggestion
              n_dot = 0;
```

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 36]
```diff
@@ -0,0 +1,885 @@
...
+            }
+          //update  n[i] = in.composition[q][c_idx_n[i]] with adding to it n_dot*dt
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]:
```suggestion
          // update  n[i] = in.composition[q][c_idx_n[i]] with adding to it n_dot*dt
```

**Comment by gassmoeller** [comment]
Agi are you done with addressing the comments?

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 37]
```diff
@@ -0,0 +1,885 @@
...
+         * variable is assigned the rank-four identity tensor by default.
+         * This leaves the isotropic constitutive law unchanged if the material
+         * model does not explicitly assign a value.
+         */
+        std::vector<SymmetricTensor<4,dim> > stress_strain_directors;
+    };
+    namespace
+    {
+      template <int dim>
+
```
[the remark this answers is not in the document]
Comment (bangerth) [reply cookbooks/anisotropic_viscosity/av_material.cc]: Please delete the empty line here (or just accept my suggestion :-) )

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 38]
```diff
@@ -0,0 +1,867 @@
...
+    template <int dim>
+    void
+    AV<dim>::evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                       MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      MaterialModel::AnisotropicViscosity<dim> *anisotropic_viscosity =
+        out.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >();
+
+      //Simple<dim>::evaluate(in,out);
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: So now I'm confused. You derive the class from the `Simple` model, but you don't actually use the `Simple` implementation anywhere, do you? Why not call the function you have commented out here? You overwrite various variables below, but I'm not clear why you need to do that.

Alternatively, if you don't want to build on the `Simple` model, then remove this line, remove the derivation of this class from `Simple`, and update the comment in front of the class that mentions `Simple`.

You'll have to do the same for the other material model.

**Comment by KiralyAgi** [comment]
I modified the plugin based on the comments. Deleted class Anisotropy (as it was not used), and deleted the line which evaluates Simple.

File: cookbooks/anisotropic_viscosity/av_material.cc [hunk 39]
```diff
@@ -0,0 +1,840 @@
...
+      std::vector<double> output(stress_strain_directors.size());
+      for (unsigned int i = 0; i < stress_strain_directors.size() ; ++i)
+        {
+          output[i]= stress_strain_directors[i][Tensor<4,dim>::unrolled_to_component_indices(idx)];
+        }
+      return output;
+    }
+  }
+}
+#endif
```
Comment (bangerth) [inline cookbooks/anisotropic_viscosity/av_material.cc]: This actually needs to move to the very bottom of the file.

Comment (KiralyAgi) [reply cookbooks/anisotropic_viscosity/av_material.cc]: I moved endif to the end of the file. I didn't experiment with squashing the patches.
Thanks for the help!

**Comment by bangerth** [comment]
Let's not worry about squashing then. We like to do it (for good reasons), but this is one of your first pull requests -- so let's go with it! :-)

We'll teach you how to squash next time around ;-)

**Pull request merged**